dds/idl/dds_visitor.cpp

```diff
@@ -233,21 +236,79 @@ dds_visitor::visit_structure(AST_Structure* node)
   ACE_UNUSED_ARG(g);
 
   // Check That Sample Keys Are Valid
+  TopicKeys topic_keys(node);
   try {
-    TopicKeys(node).count();
+    topic_keys.count();
   } catch (TopicKeys::Error& error) {
     idl_global->err()->misc_error(error.what(), error.node());
     return -1;
   }
 
-  if (idl_global->is_dcps_type(node->name())) {
+  IDL_GlobalData::DCPS_Data_Type_Info* info = idl_global->is_dcps_type(node->name());
+  if (info) {
     if (be_global->warn_about_dcps_data_type()) {
       idl_global->err()->misc_warning("\n"
         "  DCPS_DATA_TYPE and DCPS_DATA_KEY pragma statements are deprecated; please use\n"
-        "  topic annotations instead.\n"
-        "  See docs/migrating_to_topic_annotations.md in the OpenDDS source code for\n"
+        "  topic type annotations instead.\n"
+        "  See docs/migrating_to_topic_type_annotations.md in the OpenDDS source code for\n"
         "  more information.", node);
     }
+
+    /*
+     * If the struct is declared a topic type using both the older and newer
+     * styles, warn if the keys are inconsistent.
+     */
+    if (be_global->is_topic_type(node)) {
+      set<string> topic_type_keys, dcps_data_type_keys;
+
+      TopicKeys::Iterator finished = topic_keys.end();
+      for (TopicKeys::Iterator i = topic_keys.begin(); i != finished; ++i) {
+        topic_type_keys.insert(i.path());
+      }
+
+      IDL_GlobalData::DCPS_Data_Type_Info_Iter iter(info->key_list_);
+      for (ACE_TString* kp = 0; iter.next(kp) != 0; iter.advance()) {
+        dcps_data_type_keys.insert(ACE_TEXT_ALWAYS_CHAR(kp->c_str()));
+      }
+
+      if (topic_type_keys != dcps_data_type_keys) {
+        string message = "\n"
+          "  The keys are inconsistent on this struct declared to be a topic type using\n"
+          "  both a DCPS_DATA_TYPE pragma and the annotation-based system.";
+
+        bool header = false;
+        for (set<string>::iterator i = topic_type_keys.begin();
+            i != topic_type_keys.end(); ++i) {
+          if (dcps_data_type_keys.find(*i) == dcps_data_type_keys.end()) {
+            if (!header) {
+              message += "\n\n"
+                "  The following keys were declared using @key, but not DCPS_DATA_KEY:";
+              header = true;
+            }
+            message += string("\n    ") + *i;
```

Comment:
```suggestion
            message += "\n    " + *i;
```
Not a big deal but `operator+` already handles this.